enhancements/olm-deletes-operands.md

```diff
@@ -0,0 +1,294 @@
+- if `spec.cleanup.confirmationKey` is unset AND `status.cleanup.confirmationKey` is set:
+  - No cleanup, but OLM adds a status condition message to indicate that it's waiting for user confirmation
+- if `spec.cleanup.confirmationKey` == `status.cleanup.confirmationKey`
+  - OLM initiates cleanup
+- if `spec.cleanup.confirmationKey` != `status.cleanup.confirmationKey`
+  - No cleanup, and OLM adds status condition message that the cleanup confirmation failed because the keys don't match
+
+```YAML
+apiVersion: operators.coreos.com/v1alpha1
+kind: ClusterServiceVersion
+...
+spec:
+  ...
+  cleanup:
+    supported: false
+    enabled: true
+    confirmationKey: "yi7b1bC7z5"
+status:
+  ...
+  cleanup:
+    confirmationKey: "yi7b1bC7z5"
+```
+
+#### Operand cleanup status
+
+To provide visibility on the status of operand cleanup, OLM can list out the CRs that are still pending deletion in the CSV status e.g `status.cleanup.pendingDeletion`.
+
+```YAML
+apiVersion: operators.coreos.com/v1alpha1
+kind: ClusterServiceVersion
+...
+status:
+  ...
+  cleanup:
+    pendingDeletion:
+    - resource: etcdclusters.etcd.database.coreos.com
+      kind: EtcdCluster
+      version: v1beta2
+      name: Foo
+      namespace: Bar
+    - kind: EtcdCluster
+      ...
+```
+
+This information can be useful when debugging a stalled operator uninstall process to see the CRs whose removal is blocking operand cleanup.
+
+**NOTE:** The number of CRs that OLM has to list in the status is unbounded and could potentially cause the CSV size to exceed the [max request size][max-request-size] configured in etcd. For this reason there would have to be an upper limit on the number of CRs displayed in `status.cleanup.pendingDeletion`.
+
+The CSV can also be updated with a status condition to indicate that cleanup is in progress when the CSV is blocked on deletion.
+
+```YAML
+status:
+  conditions:
+  - lastTransitionTime: "..."
+    lastUpdateTime: "..."
+    message: |
+      waiting for operator to finish cleanup for 4 CRs
+    phase: Cleanup
+    reason: CleanupWaiting
+```
+
+This CSV will transition to a new phase `Cleanup` with the reason `CleanupWaiting` when it is pending deletion with the finalizer `operatorframework.io/cleanup-apis` present.
+
+#### Handling CSV deletion due to replacement
+
+A CSV can also be [deleted by OLM][csv-deletion-on-upgrade] during it's lifecycle when OLM detects a newer CSV that replaces it as part of an upgrade. For an operator that has already opted-in to operand cleanup, this would result in OLM cleaning up all the CRs before removing the old CSV.
+
+To prevent the cleanup of CRs on an upgrade, OLM should ensure that a CSV which has `spec.cleanup.enabled: true` and is marked for replacement (`phase: replacing`), should first be cleared of the `operatorframework.io/cleanup-apis` finalizer and opted-out of cleanup before it is marked for deletion or deleted.
+
+#### CRD removal
```

Comment:
what's the net here?  is this feature going to remove the CRDs or not?  It sounds like "sometimes" (which is also ok, but we need to explicilty state somewhere what is going to be implemented).

since this feature is strictly about removing operands, i'm somewhat inclined to punt on CRDs.  However if we deliver this api, we can't later update the behavior to also remove CRDs, so that's an argument for doing it now, if we think we're going to want to do it eventually.

personally i think removing the CRDs is a bit perilous.  It's possible for someone to have created their own CR instances for whatever reason, which will now be removed when the CRD is deleted, even though their CRs aren't owned by the operator.

So overall i think i gravitate towards not doing it.  Interested in hearing the arguments "for" though.

I assume if someone reinstalls an operator and the CRD already exists, that does not cause a problem (just no-ops the creation of the CRD)

Reply:
@bparees The net here is that OLM would remove the CRD if the operator that provides the API is a global/singleton install. For a single/multi namespace-scoped operator, also remove the CRD, only if it is the only operator installation in the cluster. So presumably nobody else misses that CRD when it's gone.

I can try to reword the blurb below to make that more explicit.
But I'm also actually leaning towards not removing the CRD as part of cleanup. 

I initially added it to the proposal since it was one of the overall goals for a more "complete" cleanup of all changes that came with an operator install.

However CRD removal isn't directly relevant to the CR/workload cleanup (that's more important to the user) and carries a fair amount risk with the edge cases that you've pointed out. And as a user you'd always need to reason about the operator's scope to figure out if the CRD will also be removed, so the expectation isn't as clear.

WDYT @kevinrizza

Reply:
I think we can punt on CRD removal until we implement globally scoped operators. At that point, we can always add that feature without a lot of these concerns.

Reply:
> I think we can punt on CRD removal until we implement globally scoped operators. At that point, we can always add that feature without a lot of these concerns.

+1